src/map-feature.js

```diff
@@ -2,36 +2,281 @@ export class MapFeature extends HTMLElement {
+          bboxExtent = [Infinity, Infinity, Number.NEGATIVE_INFINITY, Number.NEGATIVE_INFINITY];
+      for (let shape of shapes) {
+        let coord = shape.querySelectorAll('map-coordinates');
+        for (let i = 0; i < coord.length; ++i) {
+          _updateExtent(shape, coord[i]);
+        }
+      }
+      let topLeft = L.point(bboxExtent[0], bboxExtent[1]);
+      let bottomRight = L.point(bboxExtent[2], bboxExtent[3]);
+      let pcrsBound = M.boundsToPCRSBounds(L.bounds(topLeft, bottomRight), map.getZoom(), map.options.projection, cs);
+      return M._convertAndFormatPCRS(pcrsBound, map);
+
+      function _updateExtent(shape, coord) {
+        let data = coord.innerHTML.split(' ');
+        switch (shape.tagName) {
+          case "MAP-POINT":
+            bboxExtent = M._updateExtent(bboxExtent, +data[0], +data[1]);
+            break;
+          case "MAP-LINESTRING":
+          case "MAP-POLYGON":
+          case "MAP-MULTIPOINT":
+          case "MAP-MULTILINESTRING":
+            for (let i = 0; i < data.length; i += 2) {
+              bboxExtent = M._updateExtent(bboxExtent, +data[i], +data[i + 1]);
+            }
+            break;
+          default:
+            break;
+        }
+      }
+    }
+
+    // a .click() method that highlights the feature and show popup on map as if doing a click
+    click(event) {
+      let g = this._groupEl,
+          rect = g.getBoundingClientRect();
+      if (!event) {
+        event = new MouseEvent ("mousedown", {
+          clientX: rect.x + rect.width / 2,
+          clientY: rect.y + rect.height / 2,
+          button: 0
+        });
+      }
+      if (typeof this.onclick === 'function') {
+        this.onclick(this, event);
+        return;
+      } else {
+        let properties = this.querySelector('map-properties');
+        this.focus(event);
+        if (g.getAttribute('role') === 'link') {
+          for (let path of g.children) {
+            path.mousedown.call(this._featureLayer, event);
+            path.mouseup.call(this._featureLayer, event);
+          }
+          document.dispatchEvent(new MouseEvent('click', {
+            bubbles: false
+          }));
+        }
+        // AFTER the mousedown and mouseup events:
+        // case 1: the layer el is not re-attached to the map, the <map-feature> el is still CONNECTED
+        // case 2: the layer el is re-attached to the map; the disconnectedCallback() is invoked;
+        //         the <map-feature> el (THIS) is now DISCONNECTED, this._featureLayer is removed and a new featureLayer is created
+        if (properties && this.isConnected) {
+          if (this._featureLayer.isPopupOpen()) {
+            this._featureLayer.closePopup();
+          } else {
+            this._featureLayer.openPopup();
+          }
+        }
+      }
+    }
+    
+    focus(event) {
+      let g = this._groupEl,
```

Comment:
for the focus method is setting a focus on this (this.groupEl) enough? i.e. g.focus().

I was testing this method on a layer with multiple features, and checking if the right feature was being focused using:

`document.activeElement.shadowRoot.activeElement`

and It seemed to be returning a different feature? Will investigate further.

Reply:
Thanks! I just added g.focus() in the newest commit. However, since the current focus method also sets the CSS style of g elements, making them highlighted in the map-pane (as if the mouse were hovering over the g element), and registers an event handler (_removeFocusState) to remove the .focus() CSS style at a specific point, I'm wondering if we still need these functionalities?

Reply:
Take a look at how focus is done when using the arrow keys to go through different features:

https://github.com/Maps4HTML/Web-Map-Custom-Element/blob/main/src/mapml/features/featureGroup.js#L137-L156